Source/console-checkers/Source/GameState.cpp

```diff
@@ -0,0 +1,384 @@
+
+const std::vector<Piece>& GameState::GetRedPlayerCapturedPieces() const
+{
+	return GetCapturedPiecesForPlayerId(Identity::Red);
+}
+
+const std::vector<Piece>& GameState::GetBlackPlayerCapturedPieces() const
+{
+	return GetCapturedPiecesForPlayerId(Identity::Black);
+}
+
+std::vector<int32_t> GameState::GetBestHintIndices() const
+{
+	return m_isHintsEnabled ? m_moveDiscoveryEngine->GetBestHintIndices() : std::vector<int32_t>{};
+}
+
+void GameState::ToggleTurnPlayer()
+{
+	Identity oldTurnPlayer = m_turnPlayerIdentity;
+
+	if (m_turnPlayerIdentity == Identity::Neutral)
+	{
+		// First turn has started, red always goes first.
+		m_turnPlayerIdentity = Identity::Red;
+	}
+	else
+	{
+		m_turnPlayerIdentity = (m_turnPlayerIdentity == Identity::Red) ? Identity::Black : Identity::Red;
+	}
+	m_game->GetUIEvents().GetTurnChangedEvent().notify();
+	spdlog::info("Player turn change. old {} player new {} player", oldTurnPlayer, m_turnPlayerIdentity);
+
+	HandleTurnStart();
+}
+
+void GameState::MovePiece(const PieceMoveDescription& moveDescription)
+{
+	if (!IsMoveDescriptionValid(moveDescription))
+	{
+		return;
+	}
+
+	const Piece* sourcePiece = GetPieceFromCoord(Utility::ToGameBoardCoordFromIndex(moveDescription.sourceIndex));
+#ifdef DEBUG
+	// We should 100% have perfectly valid indices and pieces from here out. If not, I wanna know.
+	assert(sourcePiece);
+#endif
+
+	// Cache our source piece to retain context.
+	const Piece movedPiece = *sourcePiece;
+
+	const int32_t distance = Utility::GetDistanceFromSourceIndex(moveDescription.sourceIndex, moveDescription.destIndex);
+	// We can tell by the distance between source/dest if this is a simple move or a capture.
+	const bool isMove = distance == GameplaySettings::s_moveDistance;
+	const bool isCapture = distance == GameplaySettings::s_captureDistance;
+
+	if (isMove)
+	{
+		PerformMove(moveDescription, movedPiece);
+	}
+	else if (isCapture)
+	{
+		// Player has committed to moving this piece, and will be locked into its capture chain.
+		m_touchedCapturingPiece = *sourcePiece;
+		PerformCapture(moveDescription, movedPiece);
+
+		// If a capture happened we need to see if we have any more available.
+		m_moveDiscoveryEngine->ResetDiscoveredMoves();
+		PopulateAvailableMoves();
+	}
+
+	// We just moved a piece, does it need to be kinged?
+	if (ShouldKingPiece(moveDescription.destIndex, movedPiece.identity))
+	{
+		PromotePiece(moveDescription.destIndex, movedPiece);
+	}
+
+	if (CheckAndNotifyWinCondition(WinConditionReason::AllEnemyPiecesCapturedWin) ||
+		CheckAndNotifyWinCondition(WinConditionReason::GameStateViolationDraw))
+	{
+		return;
+	}
+
+	const auto& relevantCaptureList = m_touchedCapturingPiece.has_value() ?
+		m_moveDiscoveryEngine->GetAvailableCapturesForTouchedPiece() :
+		m_moveDiscoveryEngine->GetAvailableCaptures();
+
+	if (!relevantCaptureList.empty())
+	{
+		m_game->GetUIEvents().GetAdditionalPieceCaptureRequiredEvent().notify();
+		return;
+	}
+
+	ToggleTurnPlayer();
```

Comment:
No. This should not be the responsibility of a function named MovePiece. This is side effect code.

Looking further you have a problem here. The MoveCommand::Execute is being processed as the result of your CommandInput loop. You don't have anything "stepping forward the game state". This is why you don't have a good place for ToggleTurnPlayer right now. Why is your console input "driving" your program? Big big architectural smell.